File: backend-dotnet/src/InventoryWastePredictor.Api/Services/ReportService.cs
namespace InventoryWastePredictor.Api.Services;

public class ReportService
{
    private readonly IntakeService  _intakeService;
    private readonly SummaryService _summaryService;

    public ReportService(IntakeService intakeService, SummaryService summaryService)
    {
        _intakeService  = intakeService;
        _summaryService = summaryService;
    }

    private static List<string> MonthRange(int n)
    {
        var today  = DateTime.Today;
        var months = new List<string>();
        for (int i = n - 1; i >= 0; i--)
        {
            var d = today.AddMonths(-i);
            months.Add($"{d.Year:D4}-{d.Month:D2}");
        }
        return months;
    }

    public static string NextMonth(string m)
    {
        var y  = int.Parse(m[..4]);
        var mo = int.Parse(m[5..7]);
        mo++;
        if (mo > 12) { mo = 1; y++; }
        return $"{y:D4}-{mo:D2}";
    }

    private Dictionary<string, int> PurchasedForMonth(string restaurantId, string month)
    {
        var totals = new Dictionary<string, int>();
        foreach (var r in _intakeService.IntakeRecords)
        {
            if (r.RestaurantId == restaurantId && r.DeliveryDate.StartsWith(month))
                totals[r.ItemId] = totals.GetValueOrDefault(r.ItemId, 0) + r.Units;
        }
        return totals;
    }

    private Dictionary<string, (string Name, string Category)> ItemMeta(string restaurantId)
    {
        var meta = new Dictionary<string, (string, string)>();
        foreach (var r in _intakeService.IntakeRecords)
        {
            if (r.RestaurantId == restaurantId && !meta.ContainsKey(r.ItemId))
                meta[r.ItemId] = (r.Name, r.Category);
        }
        return meta;
    }

    public static double WeightedMovingAverage(List<double> values, List<double>? weights = null)
    {
        var nonZero = values.Where(v => v > 0).ToList();
        if (nonZero.Count == 0) return 0.0;
        if (nonZero.Count == 1) return nonZero[0];

        int n = values.Count;
        var w = weights ?? Enumerable.Range(1, n).Select(i => (double)i).ToList();
        double totalW   = 0;
        double weighted = 0;
        for (int i = 0; i < n; i++)
        {
            if (values[i] > 0)
            {
                totalW   += w[i];
                weighted += values[i] * w[i];
            }
        }
        return totalW == 0 ? 0.0 : Math.Round(weighted / totalW, 1);
    }

    public Dictionary<string, object> GetHistory(string restaurantId, int months = 6)
    {
        var monthLabels = MonthRange(months);
        var meta        = ItemMeta(restaurantId);

        var allItemIds         = new SortedSet<string>();
        var monthlyPurchased   = new Dictionary<string, Dictionary<string, int>>();

        foreach (var m in monthLabels)
        {
            var p = PurchasedForMonth(restaurantId, m);
            monthlyPurchased[m] = p;
            foreach (var id in p.Keys) allItemIds.Add(id);
        }

        var series = new List<Dictionary<string, object>>();
        foreach (var itemId in allItemIds)
        {
            var info = meta.TryGetValue(itemId, out var v) ? v : (itemId, "unknown");
            var purchasedList = new List<int>();
            var consumedList  = new List<int>();

            foreach (var m in monthLabels)
            {
                int p = monthlyPurchased[m].TryGetValue(itemId, out var pv) ? pv : 0;
                // Seed purchased for months with no real data using last known value
                if (p == 0 && purchasedList.Count > 0)
                    p = (int)(purchasedList[^1] * (0.95 + (Math.Abs((itemId + m).GetHashCode()) % 10) * 0.01));

                int c = _summaryService.Consumption.TryGetValue((restaurantId, itemId, m), out var cv) ? cv : 0;
                // Seed consumed for months with no real consumption
                if (c == 0 && p > 0)
                {
                    int seed = Math.Abs((itemId + m).GetHashCode()) % 20;
                    c = Math.Max(0, (int)(p * (0.75 + seed * 0.01)));
                }

                purchasedList.Add(p);
                consumedList.Add(c);
            }

            series.Add(new Dictionary<string, object>
            {
                ["item_id"]  = itemId,
                ["name"]     = info.Name,
                ["category"] = info.Category,
                ["purchased"] = purchasedList,
                ["consumed"]  = consumedList,
            });
        }

        return new Dictionary<string, object>
        {
            ["restaurant_id"] = restaurantId,
            ["months"]        = monthLabels,
            ["series"]        = series,
        };
    }

    public Dictionary<string, object> GetForecast(string restaurantId)
    {
        var history       = GetHistory(restaurantId, months: 6);
        var monthLabels   = (List<string>)history["months"];
        var currentMonth  = monthLabels[^1];
        var forecastMonth = NextMonth(currentMonth);

        var weights   = new List<double> { 1, 1, 2, 2, 3, 3 };
        var forecasts = new List<Dictionary<string, object>>();

        foreach (var s in (List<Dictionary<string, object>>)history["series"])
        {
            var consumed  = ((List<int>)s["consumed"]).Select(v => (double)v).ToList();
            var predicted = WeightedMovingAverage(consumed, weights);
            var margin    = (int)Math.Round(predicted * 0.12);
            forecasts.Add(new Dictionary<string, object>
            {
                ["item_id"]               = s["item_id"],
                ["name"]                  = s["name"],
                ["category"]              = s["category"],
                ["predicted_consumption"] = (int)predicted,
                ["confidence_low"]        = Math.Max(0, (int)predicted - margin),
                ["confidence_high"]       = (int)predicted + margin,
            });
        }

        return new Dictionary<string, object>
        {
            ["restaurant_id"]  = restaurantId,
            ["forecast_month"] = forecastMonth,
            ["forecasts"]      = forecasts,
        };
    }
}

File: src/services/report_service.py
"""
Report service — historical bought vs consumed trends and AI consumption forecast.
Uses weighted moving average for forecasting (upgradeable to Prophet/XGBoost).
"""
import hashlib
from datetime import date
from src.services.intake_service import _INTAKE_RECORDS
from src.services.summary_service import _CONSUMPTION


def _month_range(n: int) -> list[str]:
    """Return the last n months as YYYY-MM strings, oldest first."""
    today = date.today()
    months = []
    for i in range(n - 1, -1, -1):
        total_months = today.month - 1 - i
        year = today.year + total_months // 12
        month = total_months % 12 + 1
        if total_months < 0:
            year = today.year - 1 + (total_months + 12) // 12
            month = (today.month - 1 - i) % 12
            if month <= 0:
                month += 12
        months.append(f"{year:04d}-{month:02d}")
    return months


def _next_month(m: str) -> str:
    y, mo = int(m[:4]), int(m[5:7])
    mo += 1
    if mo > 12:
        mo, y = 1, y + 1
    return f"{y:04d}-{mo:02d}"


def _stable_hash(key: str) -> int:
    """Return a stable integer hash using MD5 (not affected by PYTHONHASHSEED)."""
    return int(hashlib.md5(key.encode()).hexdigest(), 16)


def _purchased_for_month(restaurant_id: str, month: str) -> dict[str, int]:
    totals: dict[str, int] = {}
    for r in _INTAKE_RECORDS:
        if r["restaurant_id"] == restaurant_id and r["delivery_date"].startswith(month):
            totals[r["item_id"]] = totals.get(r["item_id"], 0) + r["units"]
    return totals


def _item_meta(restaurant_id: str) -> dict[str, dict]:
    meta: dict[str, dict] = {}
    for r in _INTAKE_RECORDS:
        if r["restaurant_id"] == restaurant_id and r["item_id"] not in meta:
            meta[r["item_id"]] = {"name": r["name"], "category": r["category"]}
    return meta


def _weighted_moving_average(values: list[float], weights: list[float] | None = None) -> float:
    non_zero = [v for v in values if v > 0]
    if not non_zero:
        return 0.0
    if len(non_zero) == 1:
        return float(non_zero[0])
    n = len(values)
    w = weights or list(range(1, n + 1))
    total_w = sum(w[i] for i in range(n) if values[i] > 0)
    if total_w == 0:
        return 0.0
    weighted = sum(values[i] * w[i] for i in range(n) if values[i] > 0)
    return round(weighted / total_w, 1)


def get_history(restaurant_id: str, months: int = 6) -> dict:
    month_labels = _month_range(months)
    meta = _item_meta(restaurant_id)

    all_item_ids: set[str] = set()
    monthly_purchased: dict[str, dict[str, int]] = {}
    for m in month_labels:
        purchased = _purchased_for_month(restaurant_id, m)
        monthly_purchased[m] = purchased
        all_item_ids.update(purchased.keys())

    series = []
    for item_id in sorted(all_item_ids):
        info = meta.get(item_id, {"name": item_id, "category": "unknown"})
        purchased_list = []
        consumed_list = []
        for m in month_labels:
            p = monthly_purchased[m].get(item_id, 0)
            if p == 0 and purchased_list:
                p = int(purchased_list[-1] * (0.95 + _stable_hash(item_id + m) % 10 * 0.01))
            c = _CONSUMPTION.get((restaurant_id, item_id, m), 0)
            if c == 0 and p > 0:
                seed = _stable_hash(item_id + m) % 20
                c = max(0, int(p * (0.75 + seed * 0.01)))
            purchased_list.append(p)
            consumed_list.append(c)

        series.append({
            "item_id": item_id,
            "name": info["name"],
            "category": info["category"],
            "purchased": purchased_list,
            "consumed": consumed_list,
        })

    return {
        "restaurant_id": restaurant_id,
        "months": month_labels,
        "series": series,
    }


def get_forecast(restaurant_id: str) -> dict:
    history = get_history(restaurant_id, months=6)
    current_month = history["months"][-1]
    forecast_month = _next_month(current_month)

    forecasts = []
    weights = [1, 1, 2, 2, 3, 3]

    for s in history["series"]:
        consumed = s["consumed"]
        predicted = _weighted_moving_average(consumed, weights)
        margin = round(predicted * 0.12)
        forecasts.append({
            "item_id": s["item_id"],
            "name": s["name"],
            "category": s["category"],
            "predicted_consumption": int(predicted),
            "confidence_low": max(0, int(predicted) - margin),
            "confidence_high": int(predicted) + margin,
        })

    return {
        "restaurant_id": restaurant_id,
        "forecast_month": forecast_month,
        "forecasts": forecasts,
    }
